From 3f810109415de7eb6eaff248805d1b7ff1ea2ef7 Mon Sep 17 00:00:00 2001
From: Jan Moeller <Jan.Moeller@governikus.de>
Date: Thu, 21 Aug 2025 07:26:39 +0200
Subject: a11y Android: Announce changes to name/description to TalkBack

Changes are only announced for the currently focused element using the
same mechanism as a value change as Android does not offer a more
specific event type if only the contentDescription of an
AccessibilityEvent changes.

Task-number: QTBUG-139275
Pick-to: 6.8 6.9 6.10
Change-Id: Ia8c0eeeb01b46a116f0e4d3e5ab1692d9492793e
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit cdfab6d1cd962e0bc601ee9f692840c8da871f75)
---
 .../qt/android/QtAccessibilityDelegate.java          |  6 ++++++
 .../qt/android/QtAccessibilityInterface.java         |  1 +
 .../org/qtproject/qt/android/QtActivityDelegate.java |  5 +++++
 .../platforms/android/androidjniaccessibility.cpp    | 12 ++++++++++++
 .../platforms/android/androidjniaccessibility.h      |  1 +
 src/plugins/platforms/android/androidjnimain.cpp     |  6 ++++++
 src/plugins/platforms/android/androidjnimain.h       |  1 +
 .../android/qandroidplatformaccessibility.cpp        |  3 +++
 8 files changed, 35 insertions(+)

diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
index 44e90eb87486deac5c2ffcd3ba097e9014d95a8f..6c0bd5d75eccfc402035e6f5139b3614185af549 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
@@ -263,6 +263,12 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         });
     }
 
+    void notifyDescriptionOrNameChanged(int viewId, String value)
+    {
+        if (viewId == m_focusedVirtualViewId)
+            notifyValueChanged(viewId, value);
+    }
+
     void notifyAnnouncementEvent(int viewId, String message)
     {
         QtNative.runAction(() -> {
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityInterface.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityInterface.java
index 8cf08034ca4ba80d1363bfda4be2573440f13803..f7b029dd0f5a2b4570bc421dc4968dde756df112 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityInterface.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityInterface.java
@@ -10,6 +10,7 @@ interface QtAccessibilityInterface {
     default void notifyObjectFocus(int viewId) { }
     default void notifyScrolledEvent(int viewId) { }
     default void notifyValueChanged(int viewId, String value) { }
+    default void notifyDescriptionOrNameChanged(int viewId, String value) { }
     default void notifyObjectShow(int parentId) { }
     default void notifyAnnouncementEvent(int viewId, String message) { }
 }
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegate.java
index 15aaed7bfa93484320a37d55bacd5a56c6fb94e3..58d332c72812a0fbdf322c3be5dbd1c390e7e1b8 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtActivityDelegate.java
@@ -290,6 +290,11 @@ class QtActivityDelegate extends QtActivityDelegateBase
         m_accessibilityDelegate.notifyValueChanged(viewId, value);
     }
 
+    @Override public void notifyDescriptionOrNameChanged(int viewId, String value)
+    {
+        m_accessibilityDelegate.notifyDescriptionOrNameChanged(viewId, value);
+    }
+
     @Override
     public void notifyScrolledEvent(int viewId)
     {
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index 64a1065f32121e1cc7466a9b7ec07cc1c8f6a1f4..88db149ed012d9ed5c037765ad8629ed25e71157 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -149,6 +149,18 @@ namespace QtAndroidAccessibility
         QtAndroid::notifyValueChanged(accessibilityObjectId, value);
     }
 
+    // Forward declaration
+    static QString descriptionForInterface(QAccessibleInterface *iface);
+
+    void notifyDescriptionOrNameChanged(uint accessibilityObjectId)
+    {
+        QAccessibleInterface *iface = interfaceFromId(accessibilityObjectId);
+        if (iface && iface->isValid()) {
+            const QString value = descriptionForInterface(iface);
+            QtAndroid::notifyDescriptionOrNameChanged(accessibilityObjectId, value);
+        }
+    }
+
     void notifyScrolledEvent(uint accessiblityObjectId)
     {
         QtAndroid::notifyScrolledEvent(accessiblityObjectId);
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.h y/qtbase/src/plugins/platforms/android/androidjniaccessibility.h
index c42a8ddccf2d9f890a662d76b3a9cb620c5e262b..9f21b7f80ead3ba777aff2f8a005e6460c44fba0 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.h
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.h
@@ -21,6 +21,7 @@ namespace QtAndroidAccessibility
     void notifyObjectShow(uint accessibilityObjectId);
     void notifyObjectFocus(uint accessibilityObjectId);
     void notifyValueChanged(uint accessibilityObjectId);
+    void notifyDescriptionOrNameChanged(uint accessibilityObjectId);
     void notifyScrolledEvent(uint accessibilityObjectId);
     void notifyAnnouncementEvent(uint accessibilityObjectId, const QString &message);
     void createAccessibilityContextObject(QObject *parent);
diff --git x/qtbase/src/plugins/platforms/android/androidjnimain.cpp y/qtbase/src/plugins/platforms/android/androidjnimain.cpp
index 131edc5b7a8ff4ad119d8cd98f7f58236101b2c8..e2b2c9dcd93ee1c6910a6568e8fa7f2c5648e851 100644
--- x/qtbase/src/plugins/platforms/android/androidjnimain.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjnimain.cpp
@@ -235,6 +235,12 @@ namespace QtAndroid
                 "notifyValueChanged", accessibilityObjectId, value);
     }
 
+    void notifyDescriptionOrNameChanged(uint accessibilityObjectId, const QString &value)
+    {
+        m_backendRegister->callInterface<QtJniTypes::QtAccessibilityInterface, void>(
+                "notifyDescriptionOrNameChanged", accessibilityObjectId, value);
+    }
+
     void notifyScrolledEvent(uint accessibilityObjectId)
     {
         m_backendRegister->callInterface<QtJniTypes::QtAccessibilityInterface, void>(
diff --git x/qtbase/src/plugins/platforms/android/androidjnimain.h y/qtbase/src/plugins/platforms/android/androidjnimain.h
index 1031402d364ab8aca3a3a6b7a7704750abdd36cd..5e224f26c8053b292d98d4f3a71e992ce0752299 100644
--- x/qtbase/src/plugins/platforms/android/androidjnimain.h
+++ y/qtbase/src/plugins/platforms/android/androidjnimain.h
@@ -60,6 +60,7 @@ namespace QtAndroid
     void notifyObjectShow(uint parentObjectId);
     void notifyObjectFocus(uint accessibilityObjectId);
     void notifyValueChanged(uint accessibilityObjectId, jstring value);
+    void notifyDescriptionOrNameChanged(uint accessibilityObjectId, const QString &value);
     void notifyScrolledEvent(uint accessibilityObjectId);
     void notifyAnnouncementEvent(uint accessibilityObjectId, const QString &message);
 #endif
diff --git x/qtbase/src/plugins/platforms/android/qandroidplatformaccessibility.cpp y/qtbase/src/plugins/platforms/android/qandroidplatformaccessibility.cpp
index 70067c6819e5851123a9a879afb023ff602a9180..ac8b40b9a20d089473a26f347190aac0341b23c3 100644
--- x/qtbase/src/plugins/platforms/android/qandroidplatformaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/qandroidplatformaccessibility.cpp
@@ -36,6 +36,9 @@ void QAndroidPlatformAccessibility::notifyAccessibilityUpdate(QAccessibleEvent *
         QtAndroidAccessibility::notifyValueChanged(event->uniqueId());
     } else if (event->type() == QAccessible::ScrollingEnd) {
         QtAndroidAccessibility::notifyScrolledEvent(event->uniqueId());
+    } else if (event->type() == QAccessible::NameChanged
+               || event->type() == QAccessible::DescriptionChanged) {
+        QtAndroidAccessibility::notifyDescriptionOrNameChanged(event->uniqueId());
     } else if (event->type() == QAccessible::Announcement) {
         auto *announcementEvent = static_cast<QAccessibleAnnouncementEvent *>(event);
         QtAndroidAccessibility::notifyAnnouncementEvent(announcementEvent->uniqueId(),
